File: ios/Tables/ViewerTableSurface.swift
import UIKit

/// Immutable table preparation retained by one parent prose artifact. Scrolling
/// state intentionally belongs to the mounted drawing view, never this value.
struct PreparedViewerTableCell {
    let sourcePosition: Int
    let frame: CGRect
    let contentOrigin: CGPoint
    let content: PreparedProseLayout
    let sourceCellIndex: Int?
    let isHeader: Bool
    let attributesKey: String?

    var retainedBytes: Int { 96 + content.retainedBytes }
}

/// A rendering-only table. It owns grid geometry and rich cell artifacts, but
/// has no cache, lease, editor session, or mutable viewport state.
final class ViewerTableSurface {
    let identity: String
    let hostViewportWidth: CGFloat
    let style: TableStyle
    let direction: TableLayoutDirection
    let layout: TableLayoutResult
    let cells: [PreparedViewerTableCell]
    let sourceTable: FfiViewerTable?
    let sourceAttributes: [String: [String: Any]]
    let syntheticRegions: [TableRenderSyntheticRegion]
    let preparationError: ProseViewerError?

    var bounds: CGRect { CGRect(origin: .zero, size: layout.contentSize) }
    var retainedBytes: Int {
        256 + cells.reduce(0) { $0 + $1.retainedBytes }
            + (sourceTable?.cells.count ?? 0) * 16 + syntheticRegions.count * 64
            + layout.columnWidths.count * 16 + layout.rowOffsets.count * 16
            + layout.rectangles.count * 96 + layout.sourceOrder.count * 16
    }

    init(
        identity: String,
        record: TableGridRecord,
        viewportWidth: CGFloat,
        style: TableStyle,
        direction: TableLayoutDirection,
        displayScale: CGFloat = UIScreen.main.scale,
        themeDigest: String = "",
        fontEnvironmentRevision: Int = 0,
        textScale: CGFloat = 1,
        sourceTable: FfiViewerTable? = nil,
        sourceAttributes: [String: [String: Any]] = [:],
        prepareCell: (TableGridCell, CGFloat) -> PreparedProseLayout
    ) {
        self.identity = identity
        self.hostViewportWidth = viewportWidth
        self.style = style
        self.direction = direction
        self.sourceTable = sourceTable
        self.sourceAttributes = sourceAttributes
        self.syntheticRegions = sourceTable?.syntheticRegions ?? []
        let sourceCellIndexes = Dictionary(uniqueKeysWithValues: (sourceTable?.cells ?? []).enumerated().map { (Int($0.element.sourcePos), $0.offset) })
        let sourceCells = Dictionary(uniqueKeysWithValues: record.cells.map { ($0.sourcePosition, $0) })
        let measurementRecord = TableGridRecord(
            documentOwner: record.documentOwner,
            columns: record.columns,
            rows: record.rows,
            columnWidths: record.columnWidths,
            cells: record.cells.map {
                TableGridCell(sourcePosition: $0.sourcePosition, row: $0.row, column: $0.column,
                              rowspan: $0.rowspan, colspan: $0.colspan,
                              contentKey: "\($0.contentKey):\($0.sourcePosition)",
                              attachmentRevision: $0.attachmentRevision)
            },
            failure: record.failure,
            compatibilityDiagnostic: record.compatibilityDiagnostic
        )
        var prepared: [Int: PreparedProseLayout] = [:]
        var firstPreparationError: ProseViewerError?
        let canonicalScale = displayScale.isFinite && displayScale > 0 ? displayScale : 1
        let grid = TableGridLayout(displayScale: canonicalScale)
        let resolvedLayout = grid.layout(
            record: measurementRecord,
            viewportWidth: viewportWidth,
            style: style,
            direction: direction,
            themeDigest: themeDigest,
            fontEnvironmentRevision: fontEnvironmentRevision,
            textScale: textScale
        ) { measuredCell, width in
            guard let cell = sourceCells[measuredCell.sourcePosition] else { return nil }
            let content = prepareCell(cell, width)
            prepared[cell.sourcePosition] = content
            if let error = content.error, firstPreparationError == nil { firstPreparationError = error }
            return content.size.height
        }
        for cell in record.cells where prepared[cell.sourcePosition] == nil {
            guard let frame = resolvedLayout.rectangles[cell.sourcePosition] else { continue }
            let inner = max(0, frame.width - 2 * (style.cellPadding + style.borderWidth))
            let pixels = (inner * canonicalScale).rounded()
            guard pixels.isFinite, pixels >= 0, let widthPixels = Int(exactly: pixels) else { continue }
            let content = prepareCell(cell, CGFloat(widthPixels) / canonicalScale)
            if let error = content.error, firstPreparationError == nil { firstPreparationError = error }
            prepared[cell.sourcePosition] = content
        }
        layout = resolvedLayout
        preparationError = firstPreparationError
        cells = resolvedLayout.sourceOrder.compactMap { sourcePosition in
            guard let frame = resolvedLayout.rectangles[sourcePosition], let content = prepared[sourcePosition] else {
                return nil
            }
            let sourceCellIndex = sourceCellIndexes[sourcePosition]
            let sourceCell = sourceCellIndex.flatMap { sourceTable?.cells[$0] }
            let inset = style.cellPadding + style.borderWidth
            return PreparedViewerTableCell(sourcePosition: sourcePosition, frame: frame,
                                           contentOrigin: CGPoint(x: inset, y: inset), content: content,
                                           sourceCellIndex: sourceCellIndex, isHeader: sourceCell?.header ?? false,
                                           attributesKey: sourceCell?.attrsKey)
        }
    }

    init(
        identity: String,
        hostViewportWidth: CGFloat,
        style: TableStyle,
        direction: TableLayoutDirection,
        layout: TableLayoutResult,
        cells: [PreparedViewerTableCell],
        preparationError: ProseViewerError?
    ) {
        self.identity = identity
        self.hostViewportWidth = hostViewportWidth
        self.style = style
        self.direction = direction
        self.layout = layout
        self.cells = cells
        self.sourceTable = nil
        self.sourceAttributes = [:]
        self.syntheticRegions = []
        self.preparationError = preparationError
    }

    func parentImageAttachments(offset: Int, tableOrigin: CGPoint) -> [ViewerImageAttachment] {
        var nextOrdinal = offset
        var attachments: [ViewerImageAttachment] = []
        func append(_ layout: PreparedProseLayout, at origin: CGPoint) {
            var attachmentsByID: [String: Int] = [:]
            for (index, attachment) in layout.imageAttachments.enumerated() {
                attachmentsByID[attachment.id] = index
            }
            var appended = Set<Int>()
            func appendAttachment(_ index: Int) {
                guard appended.insert(index).inserted else { return }
                let attachment = layout.imageAttachments[index]
                attachments.append(ViewerImageAttachment(
                    ordinal: nextOrdinal,
                    id: attachment.id,
                    source: attachment.source,
                    bounds: attachment.bounds.offsetBy(dx: origin.x, dy: origin.y),
                    declaredSize: attachment.declaredSize
                ))
                nextOrdinal += 1
            }
            for block in layout.blocks {
                if let attachment = block.imageAttachment, let index = attachmentsByID.removeValue(forKey: attachment.id) {
                    appendAttachment(index)
                }
                guard let nested = block.tableSurface else { continue }
                let tableBounds = block.tableBounds ?? block.bounds
                append(nested, at: CGPoint(x: origin.x + tableBounds.minX, y: origin.y + tableBounds.minY))
            }
            for index in layout.imageAttachments.indices { appendAttachment(index) }
        }
        func append(_ surface: ViewerTableSurface, at origin: CGPoint) {
            for cell in surface.cells {
                append(cell.content, at: CGPoint(x: origin.x + cell.frame.minX + cell.contentOrigin.x,
                                                  y: origin.y + cell.frame.minY + cell.contentOrigin.y))
            }
        }
        append(self, at: tableOrigin)
        return attachments
    }

    func visibleCells(in viewport: CGRect, horizontalOffset: CGFloat = 0) -> [PreparedViewerTableCell] {
        let visible = viewport.offsetBy(dx: horizontalOffset, dy: 0)
        return cells.filter { $0.frame.intersects(visible) }
    }
}

/// Mutable mounted state deliberately kept outside the immutable prepared surface.
final class ViewerTablePresentationOwner {
    private static let mapRetainedBytes = 48
    private static let entryRetainedBytes = 64
    private var logicalOffsets: [String: CGFloat] = [:]

    var retainedBytes: Int {
        guard !logicalOffsets.isEmpty else { return 0 }
        return logicalOffsets.reduce(Self.mapRetainedBytes) { total, entry in
            let keyBytes = entry.key.utf16.count.multipliedReportingOverflow(by: 2)
            let entryBytes = keyBytes.overflow ? Int.max : Self.saturatingAdd(keyBytes.partialValue, Self.entryRetainedBytes)
            return Self.saturatingAdd(total, entryBytes)
        }
    }

    func logicalOffset(for surface: ViewerTableSurface) -> CGFloat {
        logicalOffsets[surface.identity] ?? 0
    }

    func setLogicalOffset(_ offset: CGFloat, for surface: ViewerTableSurface) {
        logicalOffsets[surface.identity] = clamp(offset, for: surface)
    }

    func physicalOffset(for surface: ViewerTableSurface) -> CGFloat {
        let logical = logicalOffset(for: surface)
        let maximum = maximumOffset(for: surface)
        return surface.direction == .rightToLeft ? maximum - logical : logical
    }

    private func clamp(_ offset: CGFloat, for surface: ViewerTableSurface) -> CGFloat {
        guard offset.isFinite else { return 0 }
        return min(max(0, offset), maximumOffset(for: surface))
    }

    private func maximumOffset(for surface: ViewerTableSurface) -> CGFloat {
        let value = surface.bounds.width - surface.hostViewportWidth
        return value.isFinite ? max(0, value) : 0
    }

    private static func saturatingAdd(_ left: Int, _ right: Int) -> Int {
        guard left <= Int.max - right else { return Int.max }
        return left + right
    }
}

enum ViewerTablePresentationViewport {
    case unknown
    case known(CGRect)
}

struct ViewerTablePresentedBlock {
    let layout: PreparedProseLayout
    let block: PreparedProseBlock
    let origin: CGPoint
    let clip: CGRect
}

/// One entry per immutable layout reached by the mounted traversal. Keeping
/// decorations here prevents consumers from rebuilding table transforms.
struct ViewerTablePresentedLayout {
    let layout: PreparedProseLayout
    let origin: CGPoint
    let clip: CGRect
}

struct ViewerTablePresentedCell {
    let surface: ViewerTableSurface
    let cell: PreparedViewerTableCell
    let sourcePosition: Int
    let content: PreparedProseLayout
    let bounds: CGRect
    let contentBounds: CGRect
    let clip: CGRect
}

struct ViewerTablePresentedImage {
    /// The root attachment remains the publication owner; this is only a mounted geometry projection.
    let attachment: ViewerImageAttachment
    let sourceIdentity: String
    let bounds: CGRect
    let clip: CGRect
    let layout: PreparedProseLayout
    let block: PreparedProseBlock?
}

struct ViewerTablePresentedAtom {
    let atom: PreparedProseAtomSlot
    let sourceIdentity: String
    let bounds: CGRect
    let clip: CGRect
    let layout: PreparedProseLayout
    let block: PreparedProseBlock
}

struct ViewerTablePresentedInteraction {
    let interaction: PreparedProseInteraction
    let sourceIdentity: String
    let rects: [CGRect]
    let clip: CGRect
    let layout: PreparedProseLayout
}

struct ViewerTablePresentedAccessibilityNode {
    let node: PreparedProseAccessibilityNode
    let sourceIdentity: String
    let interactionSourceIdentity: String?
    let rects: [CGRect]
    let clip: CGRect
    let layout: PreparedProseLayout
}

struct ViewerTablePresentationSnapshot {
    let layouts: [ViewerTablePresentedLayout]
    let blocks: [ViewerTablePresentedBlock]
    let cells: [ViewerTablePresentedCell]
    let mountedCells: [ViewerTablePresentedCell]
    let images: [ViewerTablePresentedImage]
    let atoms: [ViewerTablePresentedAtom]
    let interactions: [ViewerTablePresentedInteraction]
    let accessibilityNodes: [ViewerTablePresentedAccessibilityNode]
}

/// One recursive coordinate seam for drawing, rich hits, media, atoms, and accessibility.
enum ViewerTablePresentation {
    static func project(
        layout root: PreparedProseLayout,
        owner: ViewerTablePresentationOwner,
        viewport: ViewerTablePresentationViewport
    ) -> ViewerTablePresentationSnapshot {
        var layouts: [ViewerTablePresentedLayout] = []
        var blocks: [ViewerTablePresentedBlock] = []
        var cells: [ViewerTablePresentedCell] = []
        var images: [ViewerTablePresentedImage] = []
        var atoms: [ViewerTablePresentedAtom] = []
        var interactions: [ViewerTablePresentedInteraction] = []
        var accessibilityNodes: [ViewerTablePresentedAccessibilityNode] = []
        let canonicalAttachments = Dictionary(uniqueKeysWithValues: root.imageAttachments.map { ($0.id, $0) })
        var emittedImages = Set<String>()

        func transformed(_ rect: CGRect, by origin: CGPoint) -> CGRect {
            rect.offsetBy(dx: origin.x, dy: origin.y)
        }

        func appendLayout(_ layout: PreparedProseLayout, origin: CGPoint, clip: CGRect) {
            layouts.append(ViewerTablePresentedLayout(layout: layout, origin: origin, clip: clip))
            var emittedInteractions = Set<Int>()
            var emittedAccessibility = Set<Int>()
            let interactionsByBlock = Dictionary(grouping: layout.interactions.indices.compactMap { index in
                layout.interactions[index].sourceBlockIndex.map { ($0, index) }
            }, by: { $0.0 })
            let accessibilityByBlock = Dictionary(grouping: layout.accessibilityNodes.indices.compactMap { index in
                layout.accessibilityNodes[index].sourceBlockIndex.map { ($0, index) }
            }, by: { $0.0 })

            func appendInteraction(_ index: Int) {
                guard layout.interactions.indices.contains(index), emittedInteractions.insert(index).inserted else { return }
                let interaction = layout.interactions[index]
                interactions.append(ViewerTablePresentedInteraction(
                    interaction: interaction,
                    sourceIdentity: "\(layout.key.semanticKey):interaction:\(index)",
                    rects: interaction.rects.map { transformed($0, by: origin) },
                    clip: clip,
                    layout: layout
                ))
            }

            func appendAccessibility(_ index: Int) {
                guard layout.accessibilityNodes.indices.contains(index), emittedAccessibility.insert(index).inserted else { return }
                let node = layout.accessibilityNodes[index]
                let interactionIdentity = node.interactionIndex.map { "\(layout.key.semanticKey):interaction:\($0)" }
                accessibilityNodes.append(ViewerTablePresentedAccessibilityNode(
                    node: node,
                    sourceIdentity: "\(layout.key.semanticKey):accessibility:\(index)",
                    interactionSourceIdentity: interactionIdentity,
                    rects: node.rects.map { transformed($0, by: origin) },
                    clip: clip,
                    layout: layout
                ))
            }

            for (blockIndex, block) in layout.blocks.enumerated() {
                blocks.append(ViewerTablePresentedBlock(layout: layout, block: block, origin: origin, clip: clip))
                if let attachment = block.imageAttachment, emittedImages.insert(attachment.id).inserted {
                    images.append(ViewerTablePresentedImage(
                        attachment: canonicalAttachments[attachment.id] ?? attachment,
                        sourceIdentity: "\(layout.key.semanticKey):image:\(attachment.id)",
                        bounds: transformed(attachment.bounds, by: origin),
                        clip: clip,
                        layout: layout,
                        block: block
                    ))
                }
                if let atom = block.atomSlot {
                    atoms.append(ViewerTablePresentedAtom(
                        atom: atom,
                        sourceIdentity: "\(layout.key.semanticKey):atom:\(atom.nodeType):\(atom.docPos)",
                        bounds: transformed(atom.bounds, by: origin),
                        clip: clip,
                        layout: layout,
                        block: block
                    ))
                }
                for (_, index) in interactionsByBlock[blockIndex] ?? [] { appendInteraction(index) }
                for (_, index) in accessibilityByBlock[blockIndex] ?? [] { appendAccessibility(index) }
                guard let surface = block.tableSurface, let tableBounds = block.tableBounds else { continue }
                let hostOrigin = CGPoint(x: origin.x + tableBounds.minX, y: origin.y + tableBounds.minY)
                let hostWidth = min(surface.hostViewportWidth, surface.bounds.width)
                let hostClip = clip.intersection(CGRect(origin: hostOrigin, size: CGSize(width: hostWidth, height: surface.bounds.height)))
                let contentOrigin = CGPoint(x: hostOrigin.x - owner.physicalOffset(for: surface), y: hostOrigin.y)
                for cell in surface.cells {
                    let cellBounds = transformed(cell.frame, by: contentOrigin)
                    let childOrigin = CGPoint(x: cellBounds.minX + cell.contentOrigin.x, y: cellBounds.minY + cell.contentOrigin.y)
                    let contentBounds = CGRect(origin: childOrigin, size: cell.content.size)
                    let presented = ViewerTablePresentedCell(
                        surface: surface,
                        cell: cell,
                        sourcePosition: cell.sourcePosition,
                        content: cell.content,
                        bounds: cellBounds,
                        contentBounds: contentBounds,
                        clip: hostClip
                    )
                    cells.append(presented)
                    appendLayout(cell.content, origin: childOrigin, clip: hostClip.intersection(contentBounds))
                }
            }
            for attachment in layout.imageAttachments where emittedImages.insert(attachment.id).inserted {
                images.append(ViewerTablePresentedImage(
                    attachment: canonicalAttachments[attachment.id] ?? attachment,
                    sourceIdentity: "\(layout.key.semanticKey):image:\(attachment.id)",
                    bounds: transformed(attachment.bounds, by: origin),
                    clip: clip,
                    layout: layout,
                    block: nil
                ))
            }
            for index in layout.interactions.indices { appendInteraction(index) }
            for index in layout.accessibilityNodes.indices { appendAccessibility(index) }
        }

        appendLayout(root, origin: .zero, clip: .infinite)
        let mountedCells: [ViewerTablePresentedCell]
        switch viewport {
        case .unknown:
            mountedCells = cells
        case let .known(rect) where rect.width > 0 && rect.height > 0 && rect.width.isFinite && rect.height.isFinite:
            let candidate = rect.insetBy(dx: -rect.width, dy: -rect.height)
            mountedCells = cells.filter { $0.bounds.intersects(candidate) }
        case .known:
            mountedCells = []
        }
        return ViewerTablePresentationSnapshot(
            layouts: layouts,
            blocks: blocks,
            cells: cells,
            mountedCells: mountedCells,
            images: images,
            atoms: atoms,
            interactions: interactions,
            accessibilityNodes: accessibilityNodes
        )
    }
}

File: android/src/main/java/com/apollohg/editor/tables/ViewerTableSurface.kt
package com.apollohg.editor.tables

import android.graphics.Rect
import android.graphics.RectF
import com.apollohg.editor.ProseViewerError
import com.apollohg.editor.viewer.PreparedProseLayout
import com.apollohg.editor.viewer.PreparedProseBlock
import com.apollohg.editor.viewer.PreparedProseInteraction
import com.apollohg.editor.viewer.PreparedProseAccessibilityNode
import com.apollohg.editor.viewer.PreparedViewerAtom
import com.apollohg.editor.viewer.ViewerImageAttachment
import uniffi.editor_core.FfiViewerTable

internal data class PreparedViewerTableCell(
    val sourcePosition: Int,
    val frame: TableCellRect,
    val contentOrigin: Pair<Int, Int>,
    val content: PreparedProseLayout,
    val sourceCellIndex: Int?,
    val isHeader: Boolean,
    val attributesKey: String?
) {
    val retainedBytes: Long get() = 96L + content.retainedBytes
}

/** Immutable table preparation; mounted scrolling state belongs to the drawing view. */
internal class ViewerTableSurface(
    val identity: String,
    record: TableGridRecord,
    val hostViewportWidth: Float,
    val style: TableStyle,
    val isRightToLeft: Boolean,
    displayScale: Float = 1f,
    themeDigest: String = "",
    fontEnvironmentRevision: Long = 0,
    textScale: Float = 1f,
    val sourceTable: FfiViewerTable? = null,
    val sourceAttributes: Map<String, org.json.JSONObject> = emptyMap(),
    prepareCell: (TableGridCell, Float) -> PreparedProseLayout
) {
    val layout: TableLayoutResult
    val cells: List<PreparedViewerTableCell>
    val preparationError: ProseViewerError?
    val bounds: RectF get() = RectF(0f, 0f, layout.contentWidth, layout.contentHeight)
    val retainedBytes: Long get() = 256L + cells.sumOf { it.retainedBytes } +
        (sourceTable?.cells?.size ?: 0) * 16L + layout.columnWidths.size * 16L +
        layout.rowOffsets.size * 16L + layout.rectangles.size * 48L + layout.sourceOrder.size * 16L

    init {
        val scale = displayScale.takeIf { it.isFinite() && it > 0f } ?: 1f
        val indexes = sourceTable?.cells?.mapIndexed { index, cell -> cell.sourcePos.toInt() to index }?.toMap().orEmpty()
        val sourceCells = record.cells.associateBy { it.sourcePosition }
        val measurementRecord = record.copy(cells = record.cells.map { cell ->
            cell.copy(contentKey = "${cell.contentKey}:${cell.sourcePosition}")
        })
        val prepared = mutableMapOf<Int, PreparedProseLayout>()
        var error: ProseViewerError? = null
        layout = TableGridLayout(scale).layout(measurementRecord, hostViewportWidth, style, isRightToLeft, themeDigest, fontEnvironmentRevision, textScale) { measuredCell, width ->
            val cell = sourceCells[measuredCell.sourcePosition] ?: return@layout null
            prepareCell(cell, width).also { artifact ->
                prepared[cell.sourcePosition] = artifact
                if (error == null) error = artifact.error
            }.heightPx.toFloat()
        }
        record.cells.forEach { cell ->
            if (cell.sourcePosition !in prepared) {
                val frame = layout.rectangles[cell.sourcePosition] ?: return@forEach
                val width = maxOf(0f, frame.width - 2f * (style.cellPadding + style.borderWidth))
                val pixels = kotlin.math.round(width * scale)
                if (pixels.isFinite() && pixels in 0f..Int.MAX_VALUE.toFloat()) {
                    prepareCell(cell, pixels.toInt() / scale).also { artifact ->
                        prepared[cell.sourcePosition] = artifact
                        if (error == null) error = artifact.error
                    }
                }
            }
        }
        preparationError = error
        val inset = (style.cellPadding + style.borderWidth).toInt()
        cells = layout.sourceOrder.mapNotNull { position ->
            val frame = layout.rectangles[position] ?: return@mapNotNull null
            val content = prepared[position] ?: return@mapNotNull null
            val sourceIndex = indexes[position]
            val source = sourceIndex?.let { sourceTable?.cells?.get(it) }
            PreparedViewerTableCell(position, frame, inset to inset, content, sourceIndex, source?.header ?: false, source?.attrsKey)
        }
    }

    fun parentImageAttachments(offset: Int, tableOrigin: Rect): List<ViewerImageAttachment> {
        var ordinal = offset
        val result = mutableListOf<ViewerImageAttachment>()
        lateinit var appendSurface: (ViewerTableSurface, Int, Int) -> Unit
        fun append(layout: PreparedProseLayout, originX: Int, originY: Int) {
            val byId = layout.imageAttachments.associateBy { it.id }
            val emitted = mutableSetOf<String>()
            fun emit(attachment: ViewerImageAttachment) {
                if (!emitted.add(attachment.id)) return
                result += attachment.copy(ordinal = ordinal++, bounds = Rect(attachment.bounds).apply { offset(originX, originY) })
            }
            layout.blocks.forEach { block ->
                block.imageAttachment?.let { byId[it.id]?.let(::emit) }
                block.tableSurface?.let { nested ->
                    val frame = block.tableBounds ?: block.bounds
                    appendSurface(nested, originX + frame.left, originY + frame.top)
                }
            }
            layout.imageAttachments.forEach(::emit)
        }
        appendSurface = { surface, originX, originY ->
            surface.cells.forEach { cell ->
                append(cell.content, originX + cell.frame.left.toInt() + cell.contentOrigin.first,
                    originY + cell.frame.top.toInt() + cell.contentOrigin.second)
            }
        }
        appendSurface.invoke(this, tableOrigin.left, tableOrigin.top)
        return result
    }

    fun visibleCells(viewport: RectF, horizontalOffset: Float = 0f): List<PreparedViewerTableCell> {
        val left = viewport.left + horizontalOffset
        val right = viewport.right + horizontalOffset
        if (right <= left || viewport.bottom <= viewport.top) return cells
        return cells.filter { it.frame.left < right && it.frame.left + it.frame.width > left && it.frame.top < viewport.bottom && it.frame.top + it.frame.height > viewport.top }
    }
}

/** Mutable mounted state deliberately kept outside the immutable prepared surface. */
internal class ViewerTablePresentationOwner {
    companion object {
        private const val FIXED_RETAINED_BYTES = 48L
        private const val ENTRY_RETAINED_BYTES = 48L
    }

    private val logicalOffsets = mutableMapOf<String, Float>()

    val retainedBytes: Long
        get() = FIXED_RETAINED_BYTES + logicalOffsets.entries.sumOf { entry ->
            ENTRY_RETAINED_BYTES + entry.key.length.toLong() * 2L
        }

    fun logicalOffset(surface: ViewerTableSurface): Float = logicalOffsets[surface.identity] ?: 0f

    fun setLogicalOffset(offset: Float, surface: ViewerTableSurface) {
        logicalOffsets[surface.identity] = clamp(offset, surface)
    }

    fun physicalOffset(surface: ViewerTableSurface): Float {
        val logical = logicalOffset(surface)
        val maximum = maximumOffset(surface)
        return if (surface.isRightToLeft) maximum - logical else logical
    }

    private fun clamp(offset: Float, surface: ViewerTableSurface): Float =
        if (!offset.isFinite()) 0f else offset.coerceIn(0f, maximumOffset(surface))

    private fun maximumOffset(surface: ViewerTableSurface): Float =
        (surface.bounds.width() - surface.hostViewportWidth).takeIf { it.isFinite() }?.coerceAtLeast(0f) ?: 0f
}

internal sealed interface ViewerTablePresentationViewport {
    data object Unknown : ViewerTablePresentationViewport
    data class Known(val rect: Rect) : ViewerTablePresentationViewport
}

internal data class ViewerTablePresentedBlock(
    val layout: PreparedProseLayout,
    val block: PreparedProseBlock,
    val originX: Float,
    val originY: Float,
    val clip: RectF
)

/** One entry per immutable layout reached by the mounted traversal. */
internal data class ViewerTablePresentedLayout(
    val layout: PreparedProseLayout,
    val originX: Float,
    val originY: Float,
    val clip: RectF
)

internal data class ViewerTablePresentedCell(
    val surface: ViewerTableSurface,
    val cell: PreparedViewerTableCell,
    val sourcePosition: Int,
    val content: PreparedProseLayout,
    val bounds: RectF,
    val contentBounds: RectF,
    val clip: RectF
)

internal data class ViewerTablePresentedImage(
    /** The root attachment remains the publication owner; this is mounted geometry only. */
    val attachment: ViewerImageAttachment,
    val sourceIdentity: String,
    val bounds: RectF,
    val clip: RectF,
    val layout: PreparedProseLayout,
    val block: PreparedProseBlock?
)

internal data class ViewerTablePresentedAtom(
    val atom: PreparedViewerAtom,
    val sourceIdentity: String,
    val bounds: RectF,
    val clip: RectF,
    val layout: PreparedProseLayout,
    val block: PreparedProseBlock?
)

internal data class ViewerTablePresentedInteraction(
    val interaction: PreparedProseInteraction,
    val sourceIdentity: String,
    val rects: List<RectF>,
    val clip: RectF,
    val layout: PreparedProseLayout
)

internal data class ViewerTablePresentedAccessibilityNode(
    val node: PreparedProseAccessibilityNode,
    val sourceIdentity: String,
    val interactionSourceIdentity: String?,
    val bounds: RectF,
    val clip: RectF,
    val layout: PreparedProseLayout
)

internal data class ViewerTablePresentationSnapshot(
    val layouts: List<ViewerTablePresentedLayout>,
    val blocks: List<ViewerTablePresentedBlock>,
    val cells: List<ViewerTablePresentedCell>,
    val mountedCells: List<ViewerTablePresentedCell>,
    val images: List<ViewerTablePresentedImage>,
    val atoms: List<ViewerTablePresentedAtom>,
    val interactions: List<ViewerTablePresentedInteraction>,
    val accessibilityNodes: List<ViewerTablePresentedAccessibilityNode>
)

/** One recursive coordinate seam for drawing, rich hits, media, atoms, and accessibility. */
internal object ViewerTablePresentation {
    fun project(
        root: PreparedProseLayout,
        owner: ViewerTablePresentationOwner,
        viewport: ViewerTablePresentationViewport
    ): ViewerTablePresentationSnapshot {
        val layouts = mutableListOf<ViewerTablePresentedLayout>()
        val blocks = mutableListOf<ViewerTablePresentedBlock>()
        val cells = mutableListOf<ViewerTablePresentedCell>()
        val images = mutableListOf<ViewerTablePresentedImage>()
        val atoms = mutableListOf<ViewerTablePresentedAtom>()
        val interactions = mutableListOf<ViewerTablePresentedInteraction>()
        val accessibilityNodes = mutableListOf<ViewerTablePresentedAccessibilityNode>()
        val canonicalAttachments = root.imageAttachments.associateBy { it.id }
        val emittedImages = mutableSetOf<String>()

        fun shifted(rect: Rect, x: Float, y: Float) = RectF(
            rect.left + x,
            rect.top + y,
            rect.right + x,
            rect.bottom + y
        )

        fun intersects(left: RectF, right: Rect): Boolean =
            left.right > right.left && left.left < right.right && left.bottom > right.top && left.top < right.bottom

        fun intersect(left: RectF, right: RectF): RectF = RectF(
            maxOf(left.left, right.left),
            maxOf(left.top, right.top),
            minOf(left.right, right.right),
            minOf(left.bottom, right.bottom)
        )

        fun appendLayout(layout: PreparedProseLayout, originX: Float, originY: Float, clip: RectF) {
            layouts += ViewerTablePresentedLayout(layout, originX, originY, RectF(clip))
            val emittedInteractions = mutableSetOf<Int>()
            val emittedAccessibility = mutableSetOf<Int>()
            val interactionsByBlock = layout.interactions.indices.mapNotNull { index ->
                layout.interactions[index].sourceBlockIndex?.let { it to index }
            }.groupBy({ it.first }, { it.second })
            val accessibilityByBlock = layout.accessibilityNodes.indices.mapNotNull { index ->
                layout.accessibilityNodes[index].sourceBlockIndex?.let { it to index }
            }.groupBy({ it.first }, { it.second })

            fun appendInteraction(index: Int) {
                if (index !in layout.interactions.indices || !emittedInteractions.add(index)) return
                val interaction = layout.interactions[index]
                interactions += ViewerTablePresentedInteraction(
                    interaction,
                    "${layout.key.semanticKey}:interaction:$index",
                    interaction.rects.map { shifted(it, originX, originY) },
                    RectF(clip),
                    layout
                )
            }

            fun appendAccessibility(index: Int) {
                if (index !in layout.accessibilityNodes.indices || !emittedAccessibility.add(index)) return
                val node = layout.accessibilityNodes[index]
                accessibilityNodes += ViewerTablePresentedAccessibilityNode(
                    node,
                    "${layout.key.semanticKey}:accessibility:$index",
                    "${layout.key.semanticKey}:interaction:${node.interactionIndex}".takeIf { node.interactionIndex in layout.interactions.indices },
                    shifted(node.bounds, originX, originY),
                    RectF(clip),
                    layout
                )
            }

            layout.blocks.forEachIndexed { blockIndex, block ->
                blocks += ViewerTablePresentedBlock(layout, block, originX, originY, RectF(clip))
                block.imageAttachment?.let { attachment ->
                    if (emittedImages.add(attachment.id)) {
                        images += ViewerTablePresentedImage(
                            canonicalAttachments[attachment.id] ?: attachment,
                            "${layout.key.semanticKey}:image:${attachment.id}",
                            shifted(attachment.bounds, originX, originY),
                            RectF(clip),
                            layout,
                            block
                        )
                    }
                }
                layout.viewerAtoms.filter { atom ->
                    atom.bounds.left >= block.bounds.left && atom.bounds.right <= block.bounds.right &&
                        atom.bounds.top >= block.bounds.top && atom.bounds.bottom <= block.bounds.bottom
                }.forEach { atom ->
                    atoms += ViewerTablePresentedAtom(
                        atom,
                        "${layout.key.semanticKey}:atom:${atom.nodeType}:${atom.docPos}",
                        shifted(atom.bounds, originX, originY),
                        RectF(clip),
                        layout,
                        block
                    )
                }
                interactionsByBlock[blockIndex].orEmpty().forEach(::appendInteraction)
                accessibilityByBlock[blockIndex].orEmpty().forEach(::appendAccessibility)

                val surface = block.tableSurface ?: return@forEachIndexed
                val tableBounds = block.tableBounds ?: return@forEachIndexed
                val hostX = originX + tableBounds.left
                val hostY = originY + tableBounds.top
                val hostWidth = minOf(surface.hostViewportWidth, surface.bounds.width())
                val hostClip = intersect(clip, RectF(hostX, hostY, hostX + hostWidth, hostY + surface.bounds.height()))
                val contentX = hostX - owner.physicalOffset(surface)
                surface.cells.forEach { cell ->
                    val cellBounds = RectF(
                        contentX + cell.frame.left,
                        hostY + cell.frame.top,
                        contentX + cell.frame.left + cell.frame.width,
                        hostY + cell.frame.top + cell.frame.height
                    )
                    val childX = cellBounds.left + cell.contentOrigin.first
                    val childY = cellBounds.top + cell.contentOrigin.second
                    val contentBounds = RectF(childX, childY, childX + cell.content.widthPx, childY + cell.content.heightPx)
                    cells += ViewerTablePresentedCell(
                        surface,
                        cell,
                        cell.sourcePosition,
                        cell.content,
                        cellBounds,
                        contentBounds,
                        hostClip
                    )
                    appendLayout(cell.content, childX, childY, intersect(hostClip, contentBounds))
                }
            }
            layout.imageAttachments.filter { emittedImages.add(it.id) }.forEach { attachment ->
                images += ViewerTablePresentedImage(
                    canonicalAttachments[attachment.id] ?: attachment,
                    "${layout.key.semanticKey}:image:${attachment.id}",
                    shifted(attachment.bounds, originX, originY),
                    RectF(clip),
                    layout,
                    null
                )
            }
            layout.interactions.indices.forEach(::appendInteraction)
            layout.accessibilityNodes.indices.forEach(::appendAccessibility)
        }

        appendLayout(root, 0f, 0f, RectF(-Float.MAX_VALUE, -Float.MAX_VALUE, Float.MAX_VALUE, Float.MAX_VALUE))
        val mountedCells = when (viewport) {
            ViewerTablePresentationViewport.Unknown -> cells
            is ViewerTablePresentationViewport.Known -> {
                val rect = viewport.rect
                if (rect.width() <= 0 || rect.height() <= 0) emptyList() else {
                    val candidate = Rect(
                        rect.left - rect.width(),
                        rect.top - rect.height(),
                        rect.right + rect.width(),
                        rect.bottom + rect.height()
                    )
                    cells.filter { intersects(it.bounds, candidate) }
                }
            }
        }
        return ViewerTablePresentationSnapshot(layouts, blocks, cells, mountedCells, images, atoms, interactions, accessibilityNodes)
    }
}
